### PR TITLE
Improve libp2p: DHT section

### PR DESCRIPTION
**Pull request by enolal826**

### CODE REVIEW

File: docs/libp2p/libp2p-dht.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+The closeness of a key to a peer is computed by a _distance function_ (i.e. `distance(key_id, peer_id)`), which applies an XOR operation to the IDs. The XOR result is converted into an integer to _quantify_ the distance between the IDs. For example, `distance(key1, peer9) = 4` means that the distance between the `peer9` peer and the `key1` key is `4`.
+
+The mathematical explanation about the distance function is not trivial, and it is beyond the outcomes of this lecture; however, you can find more information on the topic [here](https://en.wikipedia.org/wiki/Kademlia#Academic_significance).
+
+## The Routing Table
+
+Every peer holds a list of peer IDs that are _close_. The closeness of two peers is defined by the distance function (i.e. `distance(peer1, peer2)`), in the same way as the distance between a peer and a key. Keep in mind that both peers and keys use the same ID format.
```
[the remark this answers is not in the document]
Comment (hsanjuan) [reply docs/libp2p/libp2p-dht.md:25]: It seems docs below assume the node is just storing "close" nodes, but the reality is it stores close and not so close nodes.

File: docs/libp2p/libp2p-dht.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+### Value Storage
+
+To store a value, a mapping of the form `key -> value` is created in the DHT. For every key, an ID is generated by using the same format as peer IDs.
+
+Remember that having an ID for every key in the same format as peer IDs means that we can calculate _distances_ between key IDs and peer IDs. When adding a value, the following process occurs:
+
+1. Generate an ID for the key: `key_id`.
+2. Find the _k_ closest nodes to the key ID (i.e. `FIND_NODE(key_id)`).
+3. The `key -> value` record is stored in every node returned by the `FIND_NODE` operation.
```
Comment (hsanjuan) [inline docs/libp2p/libp2p-dht.md:61]: I'm not sure this is like this. I think it probably looks more like the Value Retrieval part.

That is, FIND_NODE will return some peers, and than then you are supposed to run FIND_NODE on those peers until you arrive to peers that are close-enough to the key. Only those will accept storing the key.

File: docs/libp2p/libp2p-dht.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+The arrows represent the close nodes of every peer.
+For simplicity, only one node (`Peer 10`) contains the key.
+
+Consider the following diagram, where we select the _k_ closest node on every iteration (`k = 2`).
+
+![Value retrieval high-level flow](<../../.gitbook/assets/libp2p-dht-flow.png>)
+
+You are in `Peer 1`, and you want to find `key1`. The traversal of nodes will be:
+1. `Peer 1` has three close nodes. The `FIND_NODE(key1)` operation selects `Peer 2` and `Peer 3` as the closest nodes because they have the lowest distance to the key.
+2. From `Peer 1`, a `FINDE_NODE(key1)` request is sent to the close nodes of `Peer 2` and `Peer 3` (i.e. `Peer 5`, `Peer 6`, `Peer 7`, `Peer 8`).
```
Comment (hsanjuan) [inline docs/libp2p/libp2p-dht.md:83]:
```suggestion
2. From `Peer 1`, a `FIND_NODE(key1)` request is sent to the close nodes of `Peer 2` and `Peer 3` (i.e. `Peer 5`, `Peer 6`, `Peer 7`, `Peer 8`).
```

**Review by hsanjuan — CHANGES_REQUESTED** [review]

**Review by walkerlj0 — COMMENTED** [review]
Looks good-- We will have to decide further on where we want to cover DHT (and to what depth in what section) but this is amazing content!

**Pull request merged**